### PR TITLE
`Immutable` primitive

### PR DESCRIPTION
**Pull request by scolsen**
This PR adds a new primitive `immutable` that marks variables as immutable. Calling `set!` on an immutable variable will throw an error.

Example use:

```
鲤 (def foo 4)
鲤 (immutable foo)
鲤 (set! foo 3)
Tried to set foo but it was declared immutable. at REPL:1:7.
鲤 
```

Counts toward #791

### CODE REVIEW

File: src/Emit.hs [hunk 1]
```diff
@@ -432,19 +432,22 @@ toC toCMode (Binder meta root) = emitterSrc (execState (visit startingIndent roo
...
-                     Just varInfo = info variable
-                 --appendToSrc (addIndent indent ++ "// " ++ show (length (infoDelete varInfo)) ++ " deleters for " ++ properVariableName ++ ":\n")
-                 delete indent varInfo
-                 appendToSrc (addIndent indent ++ properVariableName ++ " = " ++ valueVar ++ "; "
-                              ++ " // " ++ show (fromMaybe (VarTy "?") (ty variable)) ++ " = " ++ show (fromMaybe (VarTy "?") (ty value))
-                              ++ "\n")
-                 return ""
+              let immutable = metaIsTrue meta "immutable"
+              in  if immutable
+                  then error (show (CannotSet variable))
```
Comment (hellerve) [inline src/Emit.hs:437]: Should we add this reason to the `CannotSet` type error!

**Comment by eriksvedang** [comment]
Very cool, we really need something like this!

Unfortunately, one thing I don't like about this implementation is that it won't work on local variables right now, since they can't carry meta data. That could probably be fixed though, and is more of a limitation of the rest of the compiler. It might be good to look into first though, since having the ability to only make certain variables immutable seems like a strange thing to have to explain to people.

Also, can't the `immutable` primitive be implemented as a macro, using `meta-set!`?

**Comment by scolsen** [comment]
> Very cool, we really need something like this!
> 
> Unfortunately, one thing I don't like about this implementation is that it won't work on local variables right now, since they can't carry meta data. That could probably be fixed though, and is more of a limitation of the rest of the compiler. It might be good to look into first though, since having the ability to only make certain variables immutable seems like a strange thing to have to explain to people.
> 
> Also, can't the `immutable` primitive be implemented as a macro, using `meta-set!`?

Ah, good point. And you're right that the primitive could be implemented as a macro!

Regarding the let bindings, I'm not sure if it's better to improve let bindings by allowing them to take metadata or if it's better to instead introduce a new XObj for immutable variables instead. Part of me feels like using the distinct XObj is slightly cleaner and would be a bit easier to handle/maintain in the long term than relying on metadata...but then we're increasing the number of compiler changes and XObjs which isn't great.

**Comment by eriksvedang** [comment]
Should we close this for now?

**Comment by eriksvedang** [comment]
(until we've got meta on local bindings, I guess..?)

**Comment by scolsen** [comment]
Yeah, I think we can close this for now!—I have to think about the implementation a bit more.

**Pull request closed without merging**